flamedisx/source.py

```diff
@@ -598,14 +622,15 @@ def trace_differential_rate(self):
             self._differential_rate,
             input_signature=input_signature)
 
-    def differential_rate(self, data_tensor=None, autograph=True, **kwargs):
+    def differential_rate(self, i_batch=tf.constant(0, dtype=fd.int_type()),
```

Comment:
Many Source and Block functions have now gained an extra `i_batch` argument, even high-level methods like `Source.differential_rate`. 

This is not very nice.. E.g. someone reading the code and seeing both data_tensor and i_batch as an argument would assume these methods take the full data tensor and slice a batch out of it. But that's not what happens. Changing so many functions also forced changes to many unrelated places.

Could you explain why you this was necessary, and did you consider and reject alternatives? I think you're using i_batch to access one or more new data-dependent tensors with bounds info -- if that's the only use, could you e.g. add the bounds info to a data_tensor column instead (and take it from the first event in a batch)? 

Creating new data-dependent tensors as attributes may also have other consequences. I can't recall the specifics (probably something related to re-initialization -- but I know we used to have data_tensor and ptensor as attributes, then decided to pass them around explicitly later. We wouldn't have done that without a good reason.

Reply:
The reason why I couldn't put these bounds in the regular data tensor is because for the ion bounds, we calculate a separate upper and lower bound for each energy in the trimmed/stepped energy spectrum being summed over (this massively improved the accuracy/speed tradeoff). As this results in a column of bounds whose entries are themselves a (1D) tensor, we can't stack this into the original data tensor with the right shape.

My first pass attempt at this (and something I just spent a while reconsidering just now) was to create this 'special tensor' of energy-dependent ion bounds and pass around batch slices of it in the same way as the data tensor. However this is problematic because a `source` does not know how many of these special tensors are going to be needed (they are created within a block), and which block of `block_source` will need to access each one...

So my conclusion was that the only sensible way is to pass around the batch number so the blocks can see it, and index into the current batch special tensors directly within the block's computation... I didn't see any obvious issues with passing them around as attributes, but I suppose I might have missed something. If you have any immediate alternative ideas I'm happy to try implementing them, but yes, this was somewhat of a last resort solution.

Reply:
Ah ok, then could you use the `array_columns` feature to let flamedisx handle these 'tensors of columns' automatically? The WIMP energy spectrum block already does this, to have a different energy spectrum for each event (to account for modulation-induced spectral shifts), see e.g. https://github.com/FlamTeam/flamedisx/blob/main/flamedisx/lxe_blocks/energy_spectrum.py#L332. A complete neutron background model will have to use this feature as well. 

It will increase the data tensor's size a bit, but since the number of bounds ~ number of energies, and adding those is managable for the WIMP source, I don't think it's a show stopper.

Reply:
Done!